manifests/init.pp

```diff
@@ -148,104 +164,232 @@
+      base           => String,
+      attribute      => String,
+      filter         => String,
+      ssl            => Enum['off', 'on', 'tls', 'ssl'],
+      tls_cacertfile => Stdlib::Absolutepath,
+      tls_cacertdir  => Optional[Stdlib::Absolutepath],
+      tls_checkpeer  => Optional[Enum['never', 'allow', 'try', 'demand', 'hard']],
+      tls_ciphers    => Optional[String],
+      tls_cert       => Optional[Stdlib::Absolutepath],
+      tls_key        => Optional[Stdlib::Absolutepath],
+      tls_randfile   => Stdlib::Absolutepath,
+    ],
+    generic => Struct[
+      debug        => Boolean,
+      module       => Variant[Enum['internal'], Stdlib::Absolutepath],
+      mapfile      => Variant[Enum['none'], Stdlib::Filesource],
+      ignorecase   => Boolean,
+      cert_item    => Enum['cn', 'subject', 'kpn', 'email', 'upn', 'uid'],
+      use_getpwent => Boolean,
+    ],
+    subject => Struct[
+      debug      => Boolean,
+      module     => Variant[Enum['internal'], Stdlib::Absolutepath],
+      mapfile    => Variant[Enum['none'], Stdlib::Filesource],
+      ignorecase => Boolean,
+    ],
+    openssh => Struct[
+      debug  => Boolean,
+      module => Variant[Enum['internal'], Stdlib::Absolutepath],
+    ],
+    opensc  => Struct[
+      debug  => Boolean,
+      module => Variant[Enum['internal'], Stdlib::Absolutepath],
+    ],
+    pwent   => Struct[
+      debug      => Boolean,
+      module     => Variant[Enum['internal'], Stdlib::Absolutepath],
+      ignorecase => Boolean,
+    ],
+    null    => Struct[
+      debug         => Boolean,
+      module        => Variant[Enum['internal'], Stdlib::Absolutepath],
+      default_match => Boolean,
+      default_user  => String,
+    ],
+    cn      => Struct[
+      debug      => Boolean,
+      module     => Variant[Enum['internal'], Stdlib::Absolutepath],
+      mapfile    => Variant[Enum['none'], Stdlib::Filesource],
+      ignorecase => Boolean,
+    ],
+    mail    => Struct[
+      debug        => Boolean,
+      module       => Variant[Enum['internal'], Stdlib::Absolutepath],
+      mapfile      => Variant[Enum['none'], Stdlib::Filesource],
+      ignorecase   => Boolean,
+      ignoredomain => Boolean,
+    ],
+    ms      => Struct[
+      debug        => Boolean,
+      module       => Variant[Enum['internal'], Stdlib::Absolutepath],
+      ignorecase   => Boolean,
+      ignoredomain => Boolean,
+      domainname   => String,
+    ],
+    krb     => Struct[
+      debug      => Boolean,
+      module     => Variant[Enum['internal'], Stdlib::Absolutepath],
+      ignorecase => Boolean,
+      mapfile    => Variant[Enum['none'], Stdlib::Filesource],
+    ],
+    uid     => Struct[
+      debug      => Boolean,
+      module     => Variant[Enum['internal'], Stdlib::Absolutepath],
+      ignorecase => Boolean,
+      mapfile    => Variant[Enum['none'], Stdlib::Filesource],
+    ]
+  ]             $mapper_options_base    = $pam_pkcs11::params::mapper_options,
```

Comment:
There is quite some duplication here. How about we create a [type alias](https://puppet.com/docs/puppet/7/lang_type_aliases.html) here?

Reply:
Yes. Same issue, one is Optional while the other one isn't. But we can just have one type and loose the enforced requirements on the other case...

Reply:
How about we create type aliases for all the inner structs? Then we can still keep the outer structs to keep the difference between optional and mandatory, but we don't have to duplicate the inner struct difinitions.

Reply:
I will try to do that. It will make things more readable, but also a bit more hidden away and indirect. We will see.

Reply:
I created aliases for the inner and outer types now.

Reply:
Thanks, that looks really nice now!